Spark performance improvements

anonymous123-code opened the pull request:
Fixes #4389 

Sparks already had a lot of the infrastructure needed, but I basically changed spark from recalculating everything every tick to being notified when stuff changes. Note that this approach is more prone to bugs, especially when more funcitonality gets added. The issue also goes a lot more into detail. I also tested using Fabric's block cache API but it didn't turn out to be worth it.

Setup profiled:
(Screenshot TBD)
10 Minutes superflat, no mob spawning, no daylight cycle (In dev)

Runtime of: `vazkii.botania.common.entity.ManaSparkEntity.tick()`
With mana running:
|                       | Fabric | Forge |
| --------------- | ------ | -------|
| Before (ms)    | 37,269 | 92,968 |
| After (ms)       | 19,736 | 46,217 |
| After/Before | 52%      | 49%      |

When empty:
|                       | Fabric | Forge |
| --------------- | ------ | -------|
| Before (ms)    |  24,147 | 50,048 |
| After (ms)       | 3,821 | 9,726 |
| After/Before | 15%          |  19%         |

I already tested a lot of stuff before rebasing (But that one was like 1.19.2), so I'll need to retest.
Tested:
- Fabric
  - [ ] Singleplayer
  - [ ] Server
- Fabric
  - [ ] Singleplayer
  - [ ] Server

anonymous123-code committed:
Initial spark perf improvements

anonymous123-code committed:
Cache getManaReceiver

anonymous123-code committed:
Fix some transfer update bugs

anonymous123-code committed:
Integration with other blocks fixes
(everything should work now)

anonymous123-code committed:
Final changes: Remove unnecessary removeTransfers

anonymous123-code committed:
Fix weird behavior when transferring low mana amounts

anonymous123-code committed:
Better Caching impl

anonymous123-code committed:
Rebase fixes and remove caching
Having done some profiling, caching doesn't seem to be worth the effort, with some of my tests without caching performing better than with

williewillus commented:
Very nice, could you go through and add inline comments explaining what's being done at a high level? I think that would help review.

Like for example, I see the state/BE params being removed from findManaReceiver, which I'm assuming is to boost perf, but would be nice to have that called out instead of me assuming.

anonymous123-code commented:
Actually I think that one was a leftover from me implementing caching. I can change it, although that pattern might be common enough that its better to just keep it...

anonymous123-code committed:
Comment, rename wasFull for better readability

anonymous123-code committed:
Format

anonymous123-code commented:
Stuff should be better now, I left the findManaReceiver refactorings in as that method already existed before the PR.

williewillus commented on Xplat/src/main/java/vazkii/botania/common/entity/ManaSparkEntity.java:
```diff
@@ -163,29 +172,25 @@ public void tick() {
+					updateTransfers();
+				}
+				if (!transfersTowardsSelfToRegister.isEmpty()) {
+					transfersTowardsSelfToRegister.remove(0).registerTransfer(this);
```
If there's a lot of elements in the list, `remove(0)` is going to be expensive by forcing an arraylist recopy. Maybe you want a Deque instead?

williewillus commented on Xplat/src/main/java/vazkii/botania/common/entity/ManaSparkEntity.java:
```diff
@@ -212,8 +223,35 @@ public void tick() {
+						transfersTowardsSelfToRegister.add(spark);
+					}
+				}
+				Collections.shuffle(transfersTowardsSelfToRegister);
```
reasoning for the shuffles that were added here?

anonymous123-code replied:
Otherwise mana pools would be loaded in a relatively consistent order, making only some fill up. This makes it so if they're reloaded, that order is shuffled at least sometimes. A similar thing was in here before,  I'm only guessing the reason why it has been implemented: https://github.com/VazkiiMods/Botania/pull/4459/files#diff-6ab7cbbb33c3d206fc8ca670958f924a1353804a7fb8f234f3308622d23723f7L172 
I can also add the transfers directly, but that would be slightly breaking with older behavior (Previosly, and in this implementation dominant sparks would only add one transfer each tick, taking some time to initialize). This would make `transfersTowardsSelfToRegister` unnecessary.

williewillus commented on Xplat/src/main/java/vazkii/botania/common/entity/ManaSparkEntity.java:
```diff
@@ -59,7 +59,11 @@ public class ManaSparkEntity extends SparkBaseEntity implements ManaSpark {
 
 	private final Set<ManaSpark> transfers = Collections.newSetFromMap(new WeakHashMap<>());
```
not strictly related to the pr, but I think renaming this to `outgoingTransfers` might make the code clearer

williewillus commented on Xplat/src/main/java/vazkii/botania/common/entity/ManaSparkEntity.java:
```diff
@@ -59,7 +59,11 @@ public class ManaSparkEntity extends SparkBaseEntity implements ManaSpark {
 	private final Set<ManaSpark> transfers = Collections.newSetFromMap(new WeakHashMap<>());
 
-	private int removeTransferants = 2;
+	private final ArrayList<ManaSpark> transfersTowardsSelfToRegister = new ArrayList<>();
```
Any risk of sparks hanging around in this list and leaking memory? I'd either use weak references or filter out `remove`d sparks every tick.

williewillus commented on Xplat/src/main/java/vazkii/botania/common/entity/ManaSparkEntity.java:
```diff
@@ -365,12 +416,24 @@ public SparkUpgradeType getUpgrade() {
 	@Override
 	public boolean areIncomingTransfersDone() {
 		if (getAttachedManaReceiver() instanceof ManaPool) {
-			return removeTransferants > 0;
```
tbh I'm not really sure what this old code did lol, do you know and can you explain why it should be removed?

anonymous123-code replied:
RemoveTransferants was the old reset logic when stuff like the upgrade or the network changed. This is now handled differently. And the only reason why it might be removed is if it is full, which is checked seperately, so I changed it to be a constant `false`

williewillus reviewed (commented):
Looks mostly good, some more questions and we should land this for the next beta

anonymous123-code committed:
Address review:
Better names
Remove at end, to save copying performance
Avoid potential memory leak when dominant is switched in and out quickly

williewillus commented:
let's get it tested in the next beta
